### PR TITLE
Fix slow dependency resolution

### PR DESCRIPTION
**Pull request by yanzhiwei147**
For master issue [#5180](https://github.com/CocoaPods/CocoaPods/issues/5180)

### CODE REVIEW

File: lib/cocoapods-core/specification.rb [hunk 1]
```diff
@@ -80,7 +81,10 @@ def ==(other)
     # @return [Fixnum] The hash value.
     #
     def hash
-      (name.hash * 53) ^ version.hash
+      if @hash_value == 0
```
Comment (segiddins) [inline lib/cocoapods-core/specification.rb]: why not `@hash ||=`

Comment (segiddins) [inline lib/cocoapods-core/specification.rb]: you'll also need to reset the hash when changing the name or the version

**Comment by segiddins** [comment]
What sort of performance improvements does this give?

**Comment by yanzhiwei147** [comment]
@segiddins I found it calculate hash value more than 10 billion when run the `pod update `(It access frequently).

**Comment by segiddins** [comment]
Right, but could you please share a benchmark or how much time these changes save?

**Comment by yanzhiwei147** [comment]
I know the total save time only, the same as this [PR](https://github.com/CocoaPods/CocoaPods/pull/5510).

**Comment by segiddins** [comment]
If it's a 4x speedup, great! Can you please add a changelog entry and address the outstanding comments? Thanks!

**Comment by yanzhiwei147** [comment]
Yes, I can add the changelog.

**Pull request closed without merging**